### ISSUE TITLE
Plot Glyph figures with vtu files (vtuPointData)

### ISSUE THREAD

**Issue by XushanLu** [marcomusy/vedo#382]
Here come the vector plot issues mainly with vtu Glyph...

I have vtu files (binary format obtained from pyevtk) which are basically made up of points with coordinates and attributes. One scalar attribute named 'E', and one vector attribute named 'direction'. With Paraview, I can create a Glyph plot by selecting the direction with the 'direction' attribute and the scalar with the 'E' attribute. 

I think with vedo, from the examples that I see, it mainly uses Arrows object to create these 3D vector plots. Is there a way to directly plot the information in the vtu file? I can certainly go back to work with my original data before they were converted to vtu, but it would be good to know whether using vtu files directly is possible or not.

A vtu file exmaple:
[extra_B_21.vtu.zip](https://github.com/marcomusy/vedo/files/6406002/extra_B_21.vtu.zip)
Figure plotted using Paraview (the red rectangle represents a conductor):
![image](https://user-images.githubusercontent.com/7824380/116709055-e1a2a800-a9aa-11eb-9457-4be0fd26b7da.png)

**Comment by marcomusy** [comment]
Hi, I just fixed a bug (there might be more) in `Glyph` .. so after

`pip install -U git+https://github.com/marcomusy/vedo.git`

try:
```python
from vedo import *

pts = UGrid("data/chddem/extra_B_21.vtu").tomesh()
pts.cmap('jet', 'E').lighting('off').ps(8)

direction = pts.getPointArray('direction') # numpy array
printInfo(direction)

p = pts.points()
arrs = Arrows(p, p+direction*1.0e+12, c='jet') # sort of manual creation

a = Arrow(endPoint=(40,0,0))
glyph = Glyph(pts, a, orientationArray='direction', colorByScalar=True, c='jet')

pts.addScalarBar()
glyph.addScalarBar()

show(pts, arrs, glyph, N=3, bg='k1', axes=8, elevation=-30)
```
![Screenshot from 2021-05-01 01-50-06](https://user-images.githubusercontent.com/32848391/116764011-d2aa0d00-aa1f-11eb-8d40-68182767c78e.png)

to be honest I'm not sure if this is what youre aiming for or/and if it's correct at all.

> Is there a way to directly plot the information in the vtu file? I can certainly go back to work with my original data before they were converted to vtu, but it would be good to know whether using vtu files directly is possible or not.

I'm not sure why you need an unstructured grid if it's just a bunch of points. A vtkPolyData would be ok too.

**Comment by CHDtdem** [comment]
Thanks for fixing the bug!

I also did something similar to what you did there
```python
direction = pts.getPointArray('direction') # numpy array
printInfo(direction)

p = pts.points()
arrs = Arrows(p, p+direction*1.0e+12, c='jet') # sort of manual creation
```
by calling `getPointArray`, but it did not go too far. I guess the key is 
`pts.cmap('jet', 'E').lighting('off').ps(8)`?

Also, would it be possible to use log scale for the color bar? The plot I got from Paraview was using log scale and it gave something nicer than a linear scalar bar.

**Comment by marcomusy** [comment]
log-scale is not implemented.. the only way I see doing it is to do it manually:
```python
E  = pts.getPointArray('E')
pts.addPointArray(log(E), "logE")
pts.cmap('jet', 'logE').lighting('off').ps(8)
```
![Screenshot from 2021-05-02 23-59-11](https://user-images.githubusercontent.com/32848391/116828991-6dcaf000-aba2-11eb-9849-045c6a96bb40.png)

**Comment by CHDtdem** [comment]
Okay, that is good to know. I can survive well without this😂 Thanks!

**Comment by XushanLu** [comment]
Hi @marcomusy , it was a year ago that this was discussed and I am wondering whether this has been implemented. While it is true that we could simply use the method you gave above to create a log color bar, the numbers are not as straightforward as the true log color bar. You would need to do the conversion in your mind to get the normal values when you see the picture. 

If this has not been implemented, how much effort would it require to get it implemented? Thanks very much.

**Comment by marcomusy** [comment]
Hi Xushan, sorry for the late reply, I gave it a few tries but was not able to make it work properly... I'll try again asap.

**Comment by XushanLu** [comment]
Hi Marco,

I am plotting something that should ideally be using the log colorbar again. So, I had a look at the code (again). I though simply doing the following in function `_cellColors` (pointcloud.py) would be enough:
```python
lut.SetScaleToLog10()
```
Of course, it is not!

So, it seems that things such as `colors.getColor` and `colors.colorMap` should be changed. 

But would it be possible to allow the user to set the `ValuesAndLabels` for the colorbar as what we could do for the axes? If so, then we can still manually convert the numbers shown in the colorbar to the true value instead of -1 for 0.1 and -2 for 0.01.

Thanks very much!

**Comment by marcomusy** [comment]
Hi Xushan, what about this (just pushed to master):
```python
from vedo import *

sph = Sphere(r=3).z(3)

z = sph.points()[:,2]
z = exp(z)*2.5 - 1.5
print("range =", z.min(), z.max())

# lut = buildLUT(
#     [
#       (1, 'r'),
#       (10, 'g'),
#       (100, 'b'),
#       (1000, 'p'),
#     ],
#     interpolate=False,
# )
# sph.cmap(lut, z).addScalarBar()

sph.cmap("jet", z, logscale=True).addScalarBar()

sb3d = addons.ScalarBar3D(sph, c='black') # this will generate a warning, but the output is still correct

show(sph, sb3d, axes=1, viewup='z')
```
![Screenshot from 2022-08-23 16-09-16](https://user-images.githubusercontent.com/32848391/186180212-f8d471e4-7184-484c-9cf1-306a2a1bda84.png)

**Comment by XushanLu** [comment]
Hi Marco,

Thanks very much for this. From your picture, it is clear that the 2D colorbar is working as expected, i.e., an even distribution of colors and values within the colorbar in log space. However, the 3D colorbar is not behaving the same way as the 2D one: the values are still distributed evenly just in the linear space. But the colors are indeed in the log space although they are not evenly distributed within the colorbar. 

I tried this with one of my own examples, and I think I get the same thing. Please have a look at the picture:

![final_mesh 1](https://user-images.githubusercontent.com/7824380/186208071-059d4c9b-fec6-46a5-9d0d-79e0223a00db.png)

**Comment by marcomusy** [comment]
OK I think I found a way to do it, just pushed a new version.. Try again!
Same code as above:

![Screenshot from 2022-08-23 19-44-32](https://user-images.githubusercontent.com/32848391/186229545-5d91ee87-0eaa-4ead-8fb1-fefb9f703c0b.png)

Also to make b1 and b2 more visible you may give the text a `depth` or even adding a `txt.box().shift([0,0,-10]).alpha(0.5)`
Quite amazing rendering!!

**Comment by XushanLu** [comment]
Hi Marco,

Thanks very much for your very quick fix! I think it's certainly getting there although there still seems to be something strange with my example: is there a rounding problem? Or are the intervals supposed to be having the same size in log space?
![final_mesh 1](https://user-images.githubusercontent.com/7824380/186271437-ceb06660-afad-4fe4-a342-6e8fadf249ca.png)

**Comment by marcomusy** [comment]
Uhm I don't understand very well your plot... your scale looks linear (?)

about the `box` thing e.g.:
```python
from vedo import *
txt = Text3D("S_3", font="Theemim").c("red5")
tbox = txt.box().scale(1.1).shift(0,0,-.01)
tbox.wireframe(False).c("yellow8"). alpha(0.5)
show(txt, tbox)
```
<img width="215" alt="Screenshot 2022-08-23 at 23 51 54" src="https://user-images.githubusercontent.com/32848391/186273314-0c4f8e8f-b8c2-4ac1-8c81-2041a91a2d3e.png">

**Comment by XushanLu** [comment]
Oh, okay, I have not put tbox into plotter! And, indeed, the above picture is more or less linear. Here is the updated one:
![final_mesh 1](https://user-images.githubusercontent.com/7824380/186276753-5682f3b1-8bfe-4e10-9df5-75be0ad020fd.png)

And for this one, it's a bit more clear that the colorbar is indeed log:
![vidi_dc_conductivity](https://user-images.githubusercontent.com/7824380/186276831-73e31390-975f-4dab-81eb-9a709e17f0db.png)

However, there is one more issue: I cannot control the number of labels using nlabels in `addScalarbar3D`. Reducing it does reduce the number of labels, but they are not strictly controlled by `nlabels`.

**Comment by marcomusy** [comment]
I think i know what happened ... precision was not enough.. try again now..

> I cannot control the number of labels using nlabels in `addScalarbar3D`. Reducing it does reduce the number of labels, but they are not strictly controlled by `nlabels`.

that is normal: you cannot control it because it depends on the range, you can rise `nlabel`until you get "enough" labels

PS, the `box` can also work for b1 and b2, just shift it along the other axis.

**Comment by XushanLu** [comment]
Well, I don't see anything changed. What precision are you talking about?
![final_mesh 1](https://user-images.githubusercontent.com/7824380/186281271-7a03a99d-7f59-46f2-8519-7b97e5307964.png)

![vidi_dc_conductivity](https://user-images.githubusercontent.com/7824380/186281355-ce0fdf75-2169-4264-aeff-fa62230c6a37.png)


> that is normal: you cannot control it because it depends on the range, you can rise nlabel until you get "enough" labels
Alright, I did get more labels if I increase but then the labels would be superimposing on each other because there are too many zeros before a number. Is there a way to control the format of the labels? (I have manually converted the units from S/m to mS/m, partly trying to avoid having too many numbers there.)
![vidi_dc_conductivity](https://user-images.githubusercontent.com/7824380/186281440-62e0999e-3e0d-40bc-9fe6-d8f0d8ef15c7.png)

**Comment by marcomusy** [comment]
I got maybe fooled by the range.. if you notice the distance from 6-4 is about the same as 4-3 , so it's not linear...

> Is there a way to control the format of the labels?

not really unless you make them smaller... you should definitely go for microS /m `"\muS/m"`.

**Comment by XushanLu** [comment]
>not really unless you make them smaller... you should definitely go for microS /m "\muS/m".

Okay, more than enough for now, but I would assume there will be times that I would like to have some sort of finer control over the format of the labels. For example, if the value range is several orders of magnitude, then this could easily be a problem. 

But, I will stop asking for now. I've already asked too much! Thanks again for all your help. I've made these pictures thousands of times more beautiful than those made with Paraview, and I am pretty happy!

**Comment by marcomusy** [comment]
No worries! What if we provide an explicit format string like `ScalarBar(..., format=":.2f")`?

**Comment by XushanLu** [comment]
I think that certainly would be something good to have. Is it within Vedo or vtk?

**Comment by XushanLu** [comment]
This picture clearly shows there are some precision issues within the labels of the colorbar:
![vidi_dc_conductivity](https://user-images.githubusercontent.com/7824380/186487295-3fca99c8-71c0-4c88-9586-3ce0c4a0222a.png)

There should't be two 0.01.

**Comment by XushanLu** [comment]
Converting to mS/m solves the problem.
![vidi_dc_conductivity](https://user-images.githubusercontent.com/7824380/186487725-ca493802-bc68-42b3-87da-3bca9189b6b5.png)

**Comment by marcomusy** [comment]
what if you set `ScalarBar3D(labelFormat=":.3f")` or `:.3g`

**Comment by XushanLu** [comment]
Did you commit? I am updating but it seems that pip install takes a very long time to finish.

**Comment by marcomusy** [comment]
I just released a new version 2022.3.0
`pip install vedo -U` 
should pick it.

**Comment by XushanLu** [comment]
Finally installed. And it works!
![vidi_dc_conductivity](https://user-images.githubusercontent.com/7824380/186496417-c63357e0-bfde-40f1-8dc3-ce18a6df5fff.png)
Thanks very much!
